btcec/schnorr/signature.go

```diff
@@ -0,0 +1,557 @@
+			SignatureSize)
+		return nil, signatureError(ecdsa_schnorr.ErrSigTooShort, str)
+	}
+	if sigLen > SignatureSize {
+		str := fmt.Sprintf("malformed signature: too long: %d > %d", sigLen,
+			SignatureSize)
+		return nil, signatureError(ecdsa_schnorr.ErrSigTooLong, str)
+	}
+
+	// The signature is validly encoded at this point, however, enforce
+	// additional restrictions to ensure r is in the range [0, p-1], and s is in
+	// the range [0, n-1] since valid Schnorr signatures are required to be in
+	// that range per spec.
+	var r btcec.FieldVal
+	if overflow := r.SetByteSlice(sig[0:32]); overflow {
+		str := "invalid signature: r >= field prime"
+		return nil, signatureError(ecdsa_schnorr.ErrSigRTooBig, str)
+	}
+	var s btcec.ModNScalar
+	if overflow := s.SetByteSlice(sig[32:64]); overflow {
+		str := "invalid signature: s >= group order"
+		return nil, signatureError(ecdsa_schnorr.ErrSigSTooBig, str)
+	}
+
+	// Return the signature.
+	return NewSignature(&r, &s), nil
+}
+
+// IsEqual compares this Signature instance to the one passed, returning true
+// if both Signatures are equivalent. A signature is equivalent to another, if
+// they both have the same scalar value for R and S.
+func (sig Signature) IsEqual(otherSig *Signature) bool {
+	return sig.r.Equals(&otherSig.r) && sig.s.Equals(&otherSig.s)
+}
+
+// schnorrVerify attempt to verify the signature for the provided hash and
+// secp256k1 public key and either returns nil if successful or a specific error
+// indicating why it failed if not successful.
+//
+// This differs from the exported Verify method in that it returns a specific
+// error to support better testing while the exported method simply returns a
+// bool indicating success or failure.
+func schnorrVerify(sig *Signature, hash []byte, pub *btcec.PublicKey) error {
+	// The algorithm for producing a BIP-340 signature is described in
+	// README.md and is reproduced here for reference:
+	//
+	// 1. Fail if m is not 32 bytes
+	// 2. P = lift_x(int(pk)).
+	// 3. r = int(sig[0:32]); fail is r >= p.
+	// 4. s = int(sig[32:64]); fail if s >= n.
+	// 5. e = int(tagged_hash("BIP0340/challenge", bytes(r) || bytes(P) || M)) mod n.
+	// 6. R = s*G - e*P
+	// 7. Fail if is_infinite(R)
+	// 8. Fail if not hash_even_y(R)
+	// 9. Fail is x(R) != r.
+	// 10. Return success iff not failure occured before reachign this
+	// point.
+
+	// Step 1.
+	//
+	// Fail if m is not 32 bytes
+	if len(hash) != scalarSize {
+		str := fmt.Sprintf("wrong size for message (got %v, want %v)",
+			len(hash), scalarSize)
+		return signatureError(ecdsa_schnorr.ErrInvalidHashLen, str)
+	}
+
+	// Before we proceed, we want to ensure that the public key we're using
+	// for verification always has an even y-coordinate. So we'll serialize
+	// it, then parse it again to esure we only proceed with points that
+	// have an even y-coordinate.
+	pubKey, err := ParsePubKey(pub.SerializeCompressed()[1:])
```

Comment:
Nit: maybe a bit of premature optimization, we can check if `pub` has an even y-coordinate and avoid this.

Reply:
Indeed, but those the x and y are private variables as is, so we need to modify the upstream package, or convert into a jacobian point, then back to affine again to check. Also see this comment (and the one below it) for context: https://github.com/btcsuite/btcd/pull/1777#issuecomment-986333157

Reply:
"but in practice we'll always parse the public keys with the new schnorr.ParsePubKey method which'll always return points with an even y coord"  The problem with parsing the pub key and using the changed variable with a different y is that the calculated r further down may be different than the given r, failing the verification. I found this problem providing an odd pk.y to this commit. 
Link to BTC's code: https://github.com/bitcoin/bitcoin/blob/3d223712d343d49ab2fc6d0fbf66b39663835192/src/secp256k1/src/modules/schnorrsig/main_impl.h#L215

Reply:
> The problem with parsing the pub key and using the changed variable with a different y is that the calculated r further down may be different than the given r, failing the verification. I found this problem providing an odd pk.y to this commit.

Yes, before I added that fragment, we failed a test vector for this exact same reason (the challenge hash differed) as warned in this fragment of the BIP:

> Note that the correctness of verification relies on the fact that lift_x always returns a point with an even Y coordinate. A hypothetical verification algorithm that treats points as public keys, and takes the point P directly as input would fail any time a point with odd Y is used. While it is possible to correct for this by negating points with odd Y coordinate before further processing, this would result in a scheme where every (message, signature) pair is valid for two public keys (a type of malleability that exists for ECDSA as well, but we don't wish to retain). We avoid these problems by treating just the X coordinate as public key.

Our verification function "treats the points as public keys" since it accepts a `*btcec.PublicKey`, rather than a `[32]byte`. If we accepted a byte slice, and made the _caller_ do the serialization (which we do here) then we'd mirror the comment there more precisely. When a caller signs, they negate their private key if the _actual_ public key happens to have an odd y coordinate. 

If you comment out this line, the test vectors fail as the challenge hashes diverge. 

Does this follow for you @gcsfred2? I want to make sure I'm not missing anything glaring here that can cause a divergence. If we want to do things more by the book, then we can just accept a byte slice, which has the effect of moving this fragment to the line before all callers attempt to verify a signature.

Reply:
Here's the test vector in question: https://github.com/bitcoin/bips/blob/master/bip-0340/test-vectors.py#L76 (same index in our code)